File: exercise/c#/week01-code-analysis/ExerciseMovieStore/MovieStore.cs
namespace ExerciseMovieStore;

public class MovieStore
{
    public Dictionary<string, Movie> AllMovies { get; set; }
    public StoreAccount Sales { get; set; }

    public MovieStore()
    {
        AllMovies = new Dictionary<string, Movie>();
        Sales = new StoreAccount();
    }

    public void BuyMovie(string customer, string id)
    {
        if (AllMovies.TryGetValue(id, out Movie movie))
        {
            if (movie.TotalCopies - movie.BorrowedCopies > 0)
            {
                movie.TotalCopies--;
                if (movie.CanSell())
                {
                    Sales.Sell(movie, customer);
                }
                else
                {
                    Console.WriteLine("Movie not for sale");
                }
            }
            else
            {
                Console.WriteLine("All copies are currently borrowed.");
            }
        }
        else
        {
            Console.WriteLine("Movie not available!");
        }
    }

    public void AddMovie(string id, string title, string director, int totalCopies, double unitPrice)
    {
        if (AllMovies.ContainsKey(id))
        {
            Console.WriteLine("Movie already exists! Updating total copies.");
            UpdateMovieCopies(id, totalCopies);
        }
        else
        {
            AllMovies.Add(id, new Movie(id, title, director, totalCopies, unitPrice));
        }
    }

    public void UpdateMovieCopies(string id, int newTotalCopies)
    {
        if (AllMovies.TryGetValue(id, out Movie movie))
        {
            if (newTotalCopies < movie.BorrowedCopies)
            {
                Console.WriteLine("Cannot reduce total copies below borrowed copies.");
            }
            else
            {
                movie.TotalCopies = newTotalCopies;
            }
        }
        else
        {
            Console.WriteLine("Movie not found!");
        }
    }

    public void RemoveMovie(string id)
    {
        if (AllMovies.ContainsKey(id))
        {
            AllMovies.Remove(id);
        }
        else
        {
            Console.WriteLine("Movie not found!");
        }
    }

    public void BorrowMovie(string id)
    {
        if (AllMovies.TryGetValue(id, out Movie movie))
        {
            if (movie.TotalCopies - movie.BorrowedCopies > 0)
            {
                movie.BorrowedCopies++;
            }
            else
            {
                Console.WriteLine("All copies are currently borrowed.");
            }
        }
        else
        {
            Console.WriteLine("Movie not available!");
        }
    }

    public void ReturnMovie(string id)
    {
        if (AllMovies.TryGetValue(id, out Movie movie))
        {
            if (movie.BorrowedCopies > 0)
            {
                movie.BorrowedCopies--;
            }
            else
            {
                Console.WriteLine("Error: No copies were borrowed.");
            }
        }
        else
        {
            Console.WriteLine("Invalid movie ID!");
        }
    }

    public void DisplayMovies()
    {
        foreach (var movie in AllMovies.Values)
        {
            Console.WriteLine($"ID: {movie.MovieID}, Title: {movie.Title}, Director: {movie.Director}, Available Copies: {movie.TotalCopies - movie.BorrowedCopies}");
        }
    }

    public List<Movie> FindMoviesByTitle(string title)
    {
        List<Movie> result = new List<Movie>();
        foreach (var movie in AllMovies.Values)
        {
            if (movie.Title.Equals(title, StringComparison.OrdinalIgnoreCase))
            {
                result.Add(movie);
            }
        }
        return result;
    }
}

File: exercise/java/week01-code-analysis/exercise/src/main/java/movie/MovieStore.java
package movie;

import java.util.*;

public class MovieStore {
    public HashMap<String, Movie> allMovies;
    public StoreAccount sales;

    public MovieStore() {
        allMovies = new HashMap<>();
        sales = new StoreAccount();
    }

    public void buyMovie(String customer, String id) {
        Movie movie = allMovies.get(id);
        if (movie != null) {
            if (movie.totalCopies - movie.borrowedCopies > 0) {
                movie.totalCopies--;
                if(movie.canSell()) {
                    sales.sell(movie, customer);
                }
                else
                    System.out.println("Movie not for sale");
            } else {
                System.out.println("All copies are currently borrowed.");
            }
        } else {
            System.out.println("Movie not available!");
        }
    }

    public void addMovie(String id, String title, String director, int totalCopies, Double unitPrice) {
        if (allMovies.containsKey(id)) {
            System.out.println("Movie already exists! Updating total copies.");
            updateMovieCopies(id, totalCopies);
        } else {
            allMovies.put(id, new Movie(id, title, director, totalCopies, unitPrice));
        }
    }

    public void updateMovieCopies(String id, int newTotalCopies) {
        if (allMovies.containsKey(id)) {
            Movie movie = allMovies.get(id);
            if (newTotalCopies < movie.borrowedCopies) {
                System.out.println("Cannot reduce total copies below borrowed copies.");
            } else {
                movie.totalCopies = newTotalCopies;
            }
        } else {
            System.out.println("Movie not found!");
        }
    }

    public void removeMovie(String id) {
        if (allMovies.containsKey(id)) {
            allMovies.remove(id);
        } else {
            System.out.println("Movie not found!");
        }
    }

    public void borrowMovie(String id) {
        Movie movie = allMovies.get(id);
        if (movie != null) {
            if (movie.totalCopies - movie.borrowedCopies > 0) {
                movie.borrowedCopies++;
            } else {
                System.out.println("All copies are currently borrowed.");
            }
        } else {
            System.out.println("Movie not available!");
        }
    }

    public void returnMovie(String id) {
        Movie movie = allMovies.get(id);
        if (movie != null) {
            if (movie.borrowedCopies > 0) {
                movie.borrowedCopies--;
            } else {
                System.out.println("Error: No copies were borrowed.");
            }
        } else {
            System.out.println("Invalid movie ID!");
        }
    }

    public void displayMovies() {
        for (Movie m : allMovies.values()) {
            System.out.println("ID: " + m.movieID + ", Title: " + m.title + ", Director: " + m.director + ", Available Copies: " + (m.totalCopies - m.borrowedCopies));
        }
    }

    public List<Movie> findMoviesByTitle(String title) {
        List<Movie> result = new ArrayList<>();
        for (Movie m : allMovies.values()) {
            if (m.title.equalsIgnoreCase(title)) {
                result.add(m);
            }
        }
        return result;
    }
}
